felix/fv/ipsets_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2023 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build fvtests
+
+package fv_test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	api "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/fv/containers"
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	client "github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+var _ = Context("_IPSets_ Tests for IPset rendering", func() {
+
+	var (
+		etcd     *containers.Container
+		tc       infrastructure.TopologyContainers
+		felixPID int
+		client   client.Interface
+		infra    infrastructure.DatastoreInfra
+		w        *workload.Workload
+	)
+
+	BeforeEach(func() {
+		topologyOptions := infrastructure.DefaultTopologyOptions()
+		topologyOptions.FelixLogSeverity = "Info"
+		topologyOptions.EnableIPv6 = false
+		logrus.SetLevel(logrus.InfoLevel)
+		tc, etcd, client, infra = infrastructure.StartSingleNodeEtcdTopology(topologyOptions)
+		felixPID = tc.Felixes[0].GetFelixPID()
+		_ = felixPID
+		w = workload.Run(tc.Felixes[0], "w", "default", "10.65.0.2", "8085", "tcp")
+	})
+
+	AfterEach(func() {
+		w.Stop()
+		tc.Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "get", "/", "--prefix", "--keys-only")
+		}
+		etcd.Stop()
+		infra.Stop()
+	})
+
+	It("should render 10000 IP sets quickly", func() {
+		// Make 1000 network sets
+		sizes := []int{100, 1, 1, 1, 2, 3, 4, 5, 10, 10, 100, 200, 1000}
+		const numSets = 10000
+		for i := 0; i < numSets; i++ {
+			ns := api.NewGlobalNetworkSet()
+			ns.Name = fmt.Sprintf("netset-%d", i)
+			ns.Labels = map[string]string{
+				"netset": fmt.Sprintf("netset-%d", i),
+			}
+			ns.Spec.Nets = generateIPv4s(sizes[i%len(sizes)])
+			_, err := client.GlobalNetworkSets().Create(context.TODO(), ns, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+		logrus.Info(">>> CREATED NetworkSets")
```

Comment:
you could use `By()` which emits `STEP:...`

Reply:
In fact a benchmark comparison would be nice.

Reply:
Running updated benchmark (only 2k IP sets) on master:
```
STEP: Creating network sets
STEP: Creating policies with selectors
STEP: Creating a workload, activating the policies
STEP: Waiting for the first IP set to be programmed...
STEP: First IP set programmed after 2.420026617s
STEP: All IP sets programmed after 13.08048597s
STEP: Deleting workload, deactivating the policies
STEP: Waiting for first IP set to be deleted...
STEP: First IP set deleted after 10.695250385s
STEP: Recreating workload... 
STEP: Waiting for all IP sets to be recreated
STEP: All IP sets programmed after 1m22.863988693s
```
and on this PR:
```
STEP: Creating network sets
STEP: Creating policies with selectors
STEP: Creating a workload, activating the policies
STEP: Waiting for the first IP set to be programmed...
STEP: First IP set programmed after 2.38753356s
STEP: All IP sets programmed after 7.079894064s
STEP: Deleting workload, deactivating the policies
STEP: Waiting for first IP set to be deleted...
STEP: First IP set deleted after 9.426498627s
STEP: Recreating workload... 
STEP: Waiting for all IP sets to be recreated
STEP: All IP sets programmed after 7.111750296s
```
Last two lines show the improvement:  **7.1s** vs **1m22s** when we have large numbers of IP sets flapping.

Reply:
Impressive!